text/0000-target-feature.md

```diff
@@ -0,0 +1,551 @@
+
+#[target_feature="+sse3"]
+fn baz() {
+  if std::cpuid::has_feature("avx") {
+    foo(); // OK: foo is not inlined into baz
+    bar(); // ERROR: cannot inline `#[always_inline]` function bar
+           // bar requires target feature "avx" but baz only provides "sse3"
+  }
+}
+```
+
+**Example 2 (ABI):**
+
+```rust
+#[target_feature="+sse2"] 
+fn foo_sse2(a: f32x8) -> f32x8 { a } // ABI: 2x 128bit registers
+
+
+#[target_feature="+avx2"] 
+fn foo_avx2(a: f32x8) -> f32x8 { // ABI: 1x 256bit register
+  foo_sse2(a) // ABI mismatch: implicit conversion or hard error (unresolved)
+}
+```
+
+## Conditional compilation: `cfg!(target_feature)`
+
+The
+[`cfg!(target_feature = "feature_name")`](https://github.com/rust-lang/rust/issues/29717) macro
+allows querying at compile-time whether a target feature is enabled in the
+current scope. It returns `true` if the feature is enabled, and `false`
+otherwise.
+
+In a function annotated with `#[target_feature = "feature_name"]` the macro
+`cfg!(target_feature = "feature_name")` _must_ expand to `true` if the generated
+code for the function uses the feature. Otherwise, the value of the macro is
+undefined. 
+
+**Example 3 (conditional compilation):**
+
+```rust 
+fn bzhi_u32(x: u32, bit_position: u32) -> u32 {
+    // Conditional compilation: both branches must be syntactically valid,
+    // but it suffices that the true branch type-checks:
+    #[cfg!(target_feature = "bmi2")] {
+        // if this code is being compiled with BMI2 support, use a BMI2 instruction:
+        unsafe { intrinsic::bmi2::bzhi(x, bit_position) }
+    } 
+    #[not(cfg!(target_feature = "bmi2"))] {
+        // otherwise, call a portable emulation of the BMI2 instruction
+        portable_emulation::bzhi(x, bit_position)
+    } 
+}
+
+fn bzhi_u64(x: u64, bit_position: u64) -> u64 {
+    // Here both branches must type-check and whether the false branch is removed
+    // or not is left up to the optimizer.
+    if cfg!(target_feature = "bmi2") {  // `cfg!` expands to `true` or `false` at compile-time
+        // if target has the BMI2 instruction set, use a BMI2 instruction:
+        unsafe { intrinsic::bmi2::bzhi(x, bit_position) }
+    } else {
+        // otherwise call an algorithm that emulates the instruction:
+        portable_emulation::bzhi(x, bit_position)
+    }
+}
+```
+
+**Example 4 (value of `cfg!` within `#[target_feature]`):**
+
+```rust
+#[target_feature = "+avx"] 
+unsafe fn foo() {
+  if cfg!(target_feature = "avx") { /* this branch is always taken */ }
+  else { /* this branch is never taken */ }
+  #[not(cfg!(target_feature = "avx"))] {
+    // this is dead code
+  }
+}
+```
+
+## (unresolved) Run-time feature detection
+
+Writing safe wrappers around `unsafe` functions annotated with
+`#[target_feature]` requires run-time feature detection is required.
+
+How to do this is an unresolved question of this RFC that must be resolved
+before stabilization. This section does not propose a solution, but explores the
```

Comment:
What must be resolved is whether we need to stabilize run-time detection alongside this RFC or not. I need to phrase this differently, it's a bit meta.